Controller/Redirect/In.php

```diff
@@ -229,34 +249,103 @@ private function redirectToCart()
     }
 
     /**
-     * Set the session message in the message container.
-     *
-     * Only handle success & error messages.
-     * Default to Warning container if none of the above
-     * Allow custom message group for the checkout page specifically.
-     *
-     * @param \Rvvup\Payments\Api\Data\ProcessOrderResultInterface $processOrderResult
-     * @param string|null $messageGroup
-     * @return void
+     * @param string $rvvupId
+     * @param Quote $quote
+     * @param string $lastTransactionId
+     * @return Redirect|ResultInterface|void
      */
-    private function setSessionMessage(
-        ProcessOrderResultInterface $processOrderResult,
-        ?string $messageGroup = null
-    ): void {
-        // If no message to display, no action.
-        if ($processOrderResult->getCustomerMessage() === null) {
-            return;
+    private function validate(string $rvvupId, Quote &$quote, string &$lastTransactionId): ?ResultInterface
+    {
+        // First validate we have a Rvvup Order ID, silently return to basket page.
+        // A standard Rvvup return should always include `rvvup-order-id` param.
+        if ($rvvupId === null) {
+            $this->logger->error('No Rvvup Order ID provided');
+            $this->checkoutSession->restoreQuote();
+            return $this->redirectToCart();
         }
 
-        switch ($processOrderResult->getResultType()) {
-            case ProcessOrderResultInterface::RESULT_TYPE_SUCCESS:
-                $this->messageManager->addSuccessMessage(__($processOrderResult->getCustomerMessage()), $messageGroup);
-                break;
-            case ProcessOrderResultInterface::RESULT_TYPE_ERROR:
-                $this->messageManager->addErrorMessage(__($processOrderResult->getCustomerMessage()), $messageGroup);
-                break;
-            default:
-                $this->messageManager->addWarningMessage(__($processOrderResult->getCustomerMessage()), $messageGroup);
+        if (!$quote->getItems()) {
+            $quote = $this->getQuoteByRvvupId($rvvupId);
+            $lastTransactionId = (string)$quote->getPayment()->getAdditionalInformation('transaction_id');
+        }
+        if (empty($quote->getId())) {
+            $this->logger->error('Missing quote for Rvvup payment', [$rvvupId, $lastTransactionId]);
+            $this->messageManager->addErrorMessage(
+                __(
+                    'An error occurred while processing your payment (ID %1). Please contact us. ' . $rvvupId
+                )
+            );
+            return $this->redirectToCart();
         }
+
+        $hash = $quote->getPayment()->getAdditionalInformation('quote_hash');
+        $quote->collectTotals();
+        $savedHash = $this->hashService->getHashForData($quote);
+        if ($hash !== $savedHash || $rvvupId !== $lastTransactionId) {
```

Comment:
I think we should split this into two different conditions.
First one: rvvupId is not same as lastTransactionId in this case we put a different message saying, the order was already placed please review the transaction before placing again or something along those lines.

Second one: for hash doesnt match, in which case we put the message for cart has changed place another order.

Reply:
We can also confirm the exact messaging